app/Spago.hs

```diff
@@ -160,41 +164,12 @@ getDep pair@(PackageName{..}, Package{ repo = Remote repo, ..} ) = do
     processWithNewCwd = (Process.shell (Text.unpack cmd))
       { Process.cwd = Just $ Text.unpack packageDir }
 
-getAllDependencies :: Config -> [(PackageName, Package)]
-getAllDependencies Config { dependencies = deps, packages = pkgs } =
-  Map.toList $ List.foldl' go Map.empty deps
-  where
-    go acc dep
-      | Map.member dep acc = acc
-      | otherwise =
-          case Map.lookup dep pkgs of
-            -- lazy error handling, user gets crash
-            Nothing -> error $ "Package " <> show dep <> " was missing from the package set."
-            Just x@(Package { dependencies = innerDeps }) -> do
-              let newAcc = List.foldl' go acc innerDeps
-              Map.insert dep x newAcc
-
--- | Fetch all dependencies into `.spago/`
-install :: Maybe Int -> [PackageName] -> IO ()
-install maybeLimit packages = do
-  -- Make sure .spago exists
-  T.mktree $ T.fromText spagoDir
-
-  -- Only call addDependencies if new packages are supplied
-  case packages of
-    [] -> pure ()
-    additional -> 
-        -- Config is loaded here, because we will change it when
-        -- adding packages, we will reload it later on
-        -- (which will have the updates)
-        Config.ensureConfig >>= flip addDependencies additional
-
-  config <- Config.ensureConfig
 
-  let deps = getAllDependencies config
+fetchPackages :: Maybe Int -> [(PackageName, Package)] -> IO ()
+fetchPackages maybeLimit deps = do
   echoStr $ "Installing " <> show (List.length deps) <> " dependencies."
   Async.withTaskGroup limit $ \taskGroup -> do
-    asyncs <- for deps $ \dep -> Async.async taskGroup $ getDep dep
+    asyncs <- for deps $ \dep -> Async.async taskGroup $ fetchPackage dep
```

Comment:
no change needed:

oh, i missed this but my guess is that a taskgroup only approximates what you truly want here, that only a batch gets worked at a time. doesn't really matter though, because forking a pool of workers to read a shared resource mvar is a pain in the ass (see various implementations of waitPool on hackage)